ntfs_forensics/src/ntfs_forensics.cpp

```diff
@@ -0,0 +1,732 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <unordered_set>
+
+#include "ntfs_forensics.h"
+
+#include <tsk/base/tsk_base_i.h>
+
+namespace trailofbits {
+
+void processAttrs(TSK_FS_FILE* fsFile, FileInfo& result);
+
+FileInfo::FileInfo() {
+  memset(this, 0, sizeof(FileInfo));
+}
+
+std::string FileInfo::getStringRep() const {
+  // for ease of debugging
+  std::stringstream output;
+
+  output << "name: \"" << this->name << "\"\n"
+         << "path: \"" << this->path << "\"\n"
+         << "parent: " << this->filename.parent.inode << ","
+         << this->filename.parent.sequence << "\n"
+         << "btime:    " << this->standard_info_times.btime << "\n"
+         << "mtime:    " << this->standard_info_times.mtime << "\n"
+         << "ctime:    " << this->standard_info_times.ctime << "\n"
+         << "atime:    " << this->standard_info_times.atime << "\n"
+         << "fn_btime: " << this->filename.file_name_times.btime << "\n"
+         << "fn_mtime: " << this->filename.file_name_times.mtime << "\n"
+         << "fn_ctime: " << this->filename.file_name_times.ctime << "\n"
+         << "fn_atime: " << this->filename.file_name_times.atime << "\n"
+         << "type: " << typeNameFromInt(this->type) << "\n"
+         << "active: " << (this->active > 0 ? "true" : "false") << "\n";
+  std::stringstream flags, fn_flags;
+  flags << "0x" << std::hex << std::setfill('0') << std::setw(8)
+        << this->flag_val << "\n";
+  fn_flags << "0x" << std::hex << std::setfill('0') << std::setw(8)
+           << this->filename.flags << "\n";
+  output << "flags: " << flags.str() << "fn_flags: " << fn_flags.str()
+
+         << "ads: " << (this->ads == 0 ? "false" : "true") << "\n"
+         << "allocated: " << this->filename.allocated_size << "\n"
+         << "size:      " << this->filename.real_size << "\n"
+         << "inode: " << this->inode << "\n"
+         << "seq: " << this->seq << "\n"
+         << "uid: " << this->uid << "\n"
+         << "gid: " << this->gid << "\n"
+         << "owner_id: " << this->owner_id << "\n"
+         << "sid: " << this->sid << "\n";
+
+  std::stringstream oid;
+  for (int i = 0; i < 16; ++i) {
+    oid << std::hex << std::setfill('0') << std::setw(2)
+        << static_cast<unsigned>(this->object_id[i]);
+  }
+
+  output << "object_id: " << oid.str() << "\n";
+  return output.str();
+}
+
+ntfs_directory_index_entry::ntfs_directory_index_entry() {
+  memset(this, 0, sizeof(ntfs_directory_index_entry));
```

Comment:
Not sure this is a good idea https://stackoverflow.com/questions/2481654/memset-for-initialization-in-c#2481668